File: src/main/java/komu/blunt/types/checker/ExpressionTypeCheckVisitor.java
package komu.blunt.types.checker;

import komu.blunt.ast.*;
import komu.blunt.objects.Symbol;
import komu.blunt.types.*;

import java.util.ArrayList;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;
import static komu.blunt.types.Kind.STAR;
import static komu.blunt.types.Type.functionType;

public final class ExpressionTypeCheckVisitor implements ASTVisitor<Assumptions, TypeCheckResult<Type>> {
    
    private final TypeChecker tc;

    public ExpressionTypeCheckVisitor(TypeChecker tc) {
        this.tc = checkNotNull(tc);
    }

    @Override
    public TypeCheckResult<Type> visit(ASTApplication application, Assumptions as) {

        TypeCheckResult<Type> te = typeCheck(application.func, as);

        TypeCheckResult<Type> tf = typeCheck(application.arg, as);

        Type t = tc.newTVar(STAR);

        tc.unify(functionType(tf.value, t), te.value);

        TypeCheckResult.Builder<Type> result = TypeCheckResult.builder();
        result.addPredicates(te.predicates);
        result.addPredicates(tf.predicates);
        return result.build(t);
    }

    @Override
    public TypeCheckResult<Type> visit(ASTConstant constant, Assumptions as) {
        return TypeCheckResult.of(constant.valueType());
    }

    @Override
    public TypeCheckResult<Type> visit(ASTLambda lambda, Assumptions as) {
        if (lambda.arguments.size() == 1) {
            Symbol arg = lambda.arguments.get(0);

            TypeVariable argumentType = tc.newTVar(Kind.STAR);

            Assumptions as2 = Assumptions.singleton(arg, argumentType.toScheme());

            TypeCheckResult<Type> result = typeCheck(lambda.body, as.join(as2));

            return TypeCheckResult.of(functionType(argumentType, result.value), result.predicates);
        } else {
            return typeCheck(lambda.rewrite(), as);
        }
    }

    @Override
    public TypeCheckResult<Type> visit(ASTLet let, Assumptions as) {
        if (let.bindings.size() != 1)
            throw new UnsupportedOperationException("multi-var let is not supported");

        TypeCheckResult.Builder<Type> result = TypeCheckResult.builder();

        Symbol arg = let.bindings.get(0).name;
        ASTExpression exp = let.bindings.get(0).expr;

        TypeCheckResult<Type> expResult = typeCheck(exp, as);
        result.addPredicates(expResult.predicates);

        Assumptions as2 = Assumptions.singleton(arg, expResult.value.toScheme());

        TypeCheckResult<Type> bodyResult = typeCheck(let.body, as.join(as2));
        result.addPredicates(bodyResult.predicates);

        return result.build(bodyResult.value);
    }

    @Override
    public TypeCheckResult<Type> visit(ASTLetRec letRec, Assumptions as) {
        BindGroup bindGroup = new BindGroup(new ArrayList<ExplicitBinding>(), letRec.bindings);

        TypeCheckResult<Assumptions> rs = tc.typeCheckBindGroup(bindGroup, as);

        return typeCheck(letRec.body, as.join(rs.value));
    }

    @Override
    public TypeCheckResult<Type> visit(ASTSequence sequence, Assumptions as) {
        List<Predicate> predicates = new ArrayList<Predicate>();

        for (ASTExpression exp : sequence.allButLast())
            predicates.addAll(typeCheck(exp, as).predicates);

        return typeCheck(sequence.last(), as).withAddedPredicates(predicates);
    }

    @Override
    public TypeCheckResult<Type> visit(ASTSet set, Assumptions as) {
        // TODO: assume sets is always correct since it's auto-generated
        return TypeCheckResult.of(Type.UNIT);
    }

    @Override
    public TypeCheckResult<Type> visit(ASTVariable variable, Assumptions as) {
        Scheme scheme = as.find(variable.var);
        Qualified<Type> inst = tc.freshInstance(scheme);
        return TypeCheckResult.of(inst.value, inst.predicates);
    }

    @Override
    public TypeCheckResult<Type> visit(ASTList list, Assumptions as) {
        // TODO: don't include node for ASTList at all
        return typeCheck(list.rewrite(), as);
    }

    private TypeCheckResult<Type> typeCheck(ASTExpression exp, Assumptions as) {
        return tc.typeCheck(exp, as);
    }

    @Override
    public TypeCheckResult<Type> visit(ASTConstructor constructor, Assumptions as) {
        ConstructorDefinition ctor = tc.findConstructor(constructor.name);

        Qualified<Type> inst = tc.freshInstance(ctor.scheme);
        return TypeCheckResult.of(inst.value, inst.predicates);
    }

    @Override
    public TypeCheckResult<Type> visit(ASTCase astCase, Assumptions as) {
        TypeCheckResult<Type> expResult = tc.typeCheck(astCase.exp, as);

        Type type = tc.newTVar();

        TypeCheckResult.Builder<Type> result = TypeCheckResult.builder();

        for (ASTAlternative alt : astCase.alternatives) {
            PatternTypeCheckResult<Type> patternResult = tc.typeCheck(alt.pattern);

            tc.unify(expResult.value, patternResult.value);
            result.addPredicates(patternResult.predicates);
            
            TypeCheckResult<Type> valueResult = tc.typeCheck(alt.value, patternResult.as.join(as));
            tc.unify(type, valueResult.value);
            result.addPredicates(valueResult.predicates);
        }

        return result.build(type);
    }
}

File: src/main/kotlin/komu/blunt/types/checker/ExpressionTypeCheckVisitor.kt
package komu.blunt.types.checker

import komu.blunt.ast.ASTExpression
import komu.blunt.ast.BindGroup
import komu.blunt.ast.ExplicitBinding
import komu.blunt.types.BasicType
import komu.blunt.types.Predicate
import komu.blunt.types.Type
import komu.blunt.types.typeFromObject
import komu.blunt.utils.init
import java.util.*

class ExpressionTypeCheckVisitor(private val tc: TypeChecker) {

    fun typeCheck(exp: ASTExpression, ctx: Assumptions): TypeCheckResult<Type> = when (exp) {
        is ASTExpression.Application -> typeCheckApplication(exp, ctx)
        is ASTExpression.Case        -> typeCheckCase(exp, ctx)
        is ASTExpression.Constant    -> typeCheckConstant(exp)
        is ASTExpression.Constructor -> typeCheckConstructor(exp)
        is ASTExpression.Lambda      -> typeCheckLambda(exp, ctx)
        is ASTExpression.Let         -> typeCheckLet(exp, ctx)
        is ASTExpression.LetRec      -> typeCheckLetRec(exp, ctx)
        is ASTExpression.Sequence    -> typeCheckSequence(exp, ctx)
        is ASTExpression.Set         -> typeCheckSet(exp, ctx)
        is ASTExpression.Variable    -> typeCheckVariable(exp, ctx)
    }

    private fun typeCheckApplication(application: ASTExpression.Application, ass: Assumptions): TypeCheckResult<Type> {
        val type = tc.newTVar()
        val tf = typeCheck(application.func, ass)
        val ta = typeCheck(application.arg, ass);

        tc.unify(tf.value, Type.function(ta.value, type))

        return TypeCheckResult(type, tf.predicates + ta.predicates)
    }

    private fun typeCheckConstant(constant: ASTExpression.Constant): TypeCheckResult<Type> =
        TypeCheckResult(typeFromObject(constant.value))

    private fun typeCheckLambda(lambda: ASTExpression.Lambda, ass: Assumptions): TypeCheckResult<Type> {
        val argumentType = tc.newTVar()
        val (resultType, predicates) = typeCheck(lambda.body, ass.augment(lambda.argument, argumentType.toScheme()))

        return TypeCheckResult(Type.function(argumentType, resultType), predicates)
    }

    private fun typeCheckLet(let: ASTExpression.Let, ass: Assumptions): TypeCheckResult<Type> {
        val (arg, exp) = let.bindings.singleOrNull() ?: throw UnsupportedOperationException("multi-var let is not supported")

        val expResult = typeCheck(exp, ass)

        val bodyResult = typeCheck(let.body, ass.augment(arg, expResult.value.toScheme()))

        return TypeCheckResult(bodyResult.value, expResult.predicates + bodyResult.predicates)
    }

    private fun typeCheckLetRec(letRec: ASTExpression.LetRec, ass: Assumptions): TypeCheckResult<Type> {
        val bindGroup = BindGroup(ArrayList<ExplicitBinding>(), letRec.bindings)

        val rs = tc.typeCheckBindGroup(bindGroup, ass)

        return typeCheck(letRec.body, ass + rs.value)
    }

    private fun typeCheckSequence(sequence: ASTExpression.Sequence, ass: Assumptions): TypeCheckResult<Type> {
        val predicates = sequence.exps.init.flatMap { typeCheck(it, ass).predicates }

        return typeCheck(sequence.exps.last(), ass) + predicates
    }

    private fun typeCheckSet(set: ASTExpression.Set, ass: Assumptions): TypeCheckResult<Type> =
        //// TODO: assume sets is always correct since it's auto-generated
        TypeCheckResult(BasicType.UNIT)

    private fun typeCheckVariable(variable: ASTExpression.Variable, ass: Assumptions): TypeCheckResult<Type> {
        val scheme = ass[variable.name]
        val inst = tc.freshInstance(scheme)
        return TypeCheckResult(inst.value, inst.predicates)
    }

    private fun typeCheckConstructor(constructor: ASTExpression.Constructor): TypeCheckResult<Type> {
        val ctor = tc.findConstructor(constructor.name)
        val inst = tc.freshInstance(ctor.scheme)
        return TypeCheckResult(inst.value, inst.predicates)
    }

    private fun typeCheckCase(case: ASTExpression.Case, ass: Assumptions): TypeCheckResult<Type> {
        val expResult = typeCheck(case.exp, ass)

        val type = tc.newTVar()
        val predicates = ArrayList<Predicate>()

        for (alt in case.alternatives) {
            val patternResult = tc.typeCheck(alt.pattern)

            tc.unify(expResult.value, patternResult.value)
            predicates += patternResult.predicates

            val valueResult = tc.typeCheck(alt.value, patternResult.ass + ass)
            tc.unify(type, valueResult.value)
            predicates += valueResult.predicates
        }

        return TypeCheckResult(type, predicates)
    }
}
